kms/signer.go

```diff
@@ -89,56 +92,43 @@ var ErrBadDigestLength error = errors.New("specified digest length does not matc
 
 // Signer interface represents signing operations
 type Signer interface {
-	// This function continues a multiple-part signature operation, processing another data part.
+	// This function continues a multiple-part signature operation, processing
+	// another data part.
 	Update(ctx context.Context, data []byte) error
 
-	// This function finishes a single or multiple-part signature operation, possibly processing the last data part, and returns the signature.
+	// This function finishes a single or multiple-part signature operation,
+	// processing the last data part from not-nil, and returns the signature.
 	Close(ctx context.Context, data []byte) (signature []byte, err error)
 }
 
-// NewDigestSigner is a local signer which allows incremental computation of
-// the hash locally, when the underlying signature algorithm supports it. If
-// an algorithm doesn't, SignerParameters.Algorithm.Hash() will return nil.
-func NewDigestSigner(ctx context.Context, factory SignerFactory, signerParams *SignerParameters) (Signer, error) {
-	hasher := signerParams.Algorithm.Hash()
-	if hasher == nil {
-		return nil, fmt.Errorf("%w: %v", ErrUnknownDigestAlgorithm, signerParams.Algorithm.String())
-	}
-
-	return &signer{factory: factory, params: signerParams, hash: hasher}, nil
-}
-
-type signer struct {
-	factory SignerFactory
-	params  *SignerParameters
-
-	hash hash.Hash
-}
-
-func (s *signer) Update(ctx context.Context, data []byte) error {
-	_, err := s.hash.Write(data)
-	return err
-}
-
-func (s *signer) Close(ctx context.Context, data []byte) ([]byte, error) {
-	if err := s.Update(ctx, data); err != nil {
-		return nil, err
-	}
-
-	return s.factory.DigestSign(ctx, s.params, s.hash.Sum(nil))
+// DirectSignerFactory can be implemented by KMS providers which support direct
+// signing over provided hashes. This differs from pre-hashed in that the hash
+// algorithm OID is embedded in the signed payload.
+//
+// This can be optionally implemented by providers.
+type DirectSignerFactory interface {
+	// DigestSign performs a one-shot digital signatures, using a private key,
+	// from a provided digest when the algorithm supports client-side signing.
+	//
+	// SignerParameters may be mutated by the underlying provider.
+	//
+	// If the specified algorithm does not support client-side hashing, such
+	// as in the case of Ed25519 due to requiring prehash, digest may be the
+	// full message.
+	DigestSign(ctx context.Context, signerParams *SignerParameters, digest []byte) ([]byte, error)
```

Comment:
Similarly, I think it makes sense to distinguish via one of the suggestions here as `NewDigestSigner` and `DigestSign` now have different assumptions: `NewDigestSigner` will always produce a hash, but `DigestSign` is context-specific.

```suggestion
	DirectSign(ctx context.Context, signerParams *SignerParameters, digest []byte) ([]byte, error)
```

```suggestion
	Sign(ctx context.Context, signerParams *SignerParameters, digest []byte) ([]byte, error)
```